// node-dataengine-client/node-dataengine-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var de_1 = require("./libs/de");
exports.DataEngine = de_1.DataEngine;
exports.DataEngineSettings = de_1.DataEngineSettings;
exports.RequestOptions = de_1.RequestOptions;
var de_msg_competing_consumers_1 = require("./libs/de.msg.competing-consumers");
var de_msg_request_reply_1 = require("./libs/de.msg.request-reply");
var de_msg_competing_consumers_2 = require("./libs/de.msg.competing-consumers");
var Messaging;
(function (Messaging) {
    var RequestReply;
    (function (RequestReply) {
        RequestReply.Requestor = de_msg_request_reply_1.Requestor;
        RequestReply.Replier = de_msg_request_reply_1.Replier;
    })(RequestReply = Messaging.RequestReply || (Messaging.RequestReply = {}));
    ;
    var CompetingConsumers;
    (function (CompetingConsumers) {
        CompetingConsumers.CompetingConsumer = de_msg_competing_consumers_1.CompetingConsumer;
        CompetingConsumers.Producer = de_msg_competing_consumers_1.Producer;
    })(CompetingConsumers = Messaging.CompetingConsumers || (Messaging.CompetingConsumers = {}));
    ;
    var Errors;
    (function (Errors) {
        Errors.JOB_NOT_SUPPORTED = de_msg_competing_consumers_2.JOB_NOT_SUPPORTED;
        Errors.NO_WORKER_AVAILABLE = de_msg_competing_consumers_2.NO_WORKER_AVAILABLE;
    })(Errors = Messaging.Errors || (Messaging.Errors = {}));
    ;
})(Messaging = exports.Messaging || (exports.Messaging = {}));
;
var de_msg_request_reply_2 = require("./libs/de.msg.request-reply");
exports.Requestor = de_msg_request_reply_2.Requestor;
exports.Replier = de_msg_request_reply_2.Replier;
